src/d/actor/d_a_obj_hat.cpp

```diff
@@ -4,91 +4,238 @@
 bool daObjHat_c::_draw() {
-    /* Nonmatching */
+    g_env_light.settingTevStruct(TEV_TYPE_ACTOR, &this->current.pos, &this->tevStr);
+    g_env_light.setLightTevColorType(this->model, &this->tevStr);
+    this->mpMorf->updateDL();
+    return TRUE;
 }
 
+static daObjHat_c::MoveFunc_t moveProc[] = {
+    &daObjHat_c::executeNormal,
+};
+
+
 /* 0000073C-00000884       .text _execute__10daObjHat_cFv */
 bool daObjHat_c::_execute() {
-    /* Nonmatching */
+    // dCcD_Sph::~dCcD_Sp
+
+    (this->*moveProc[this->procid])();
+    f32 xspeed = this->angle.x * this->speedF;
+    f32 ymove = this->speed.y + this->gravity;
+    f32 zspeed = this->angle.z * this->speedF;
+    // because downwards vector quantity has negative value
+    if (ymove < this->maxFallSpeed) {
+        ymove = this->maxFallSpeed;
+    }
+
+    this->speed.x = xspeed;
+    this->speed.y = ymove;
+    this->speed.z = zspeed;
+    fopAcM_posMove(this, this->stts.GetCCMoveP());
+    this->acch.CrrPos(g_dComIfG_gameInfo.play.mBgS);
+    if ((this->acch.m_flags & dBgS_Acch::WALL_HIT) != 0) {
+        this->speedF = 0;
+    }
+    if ((this->acch.m_flags & dBgS_Acch::GROUND_HIT) != 0) {
+        cLib_addCalc(&this->speedF, 0.0, 0.3, 1000.0, 1.0);
+    }
+
+    this->cyl.SetC(this->current.pos);
+    g_dComIfG_gameInfo.play.mCcS.Set(&this->cyl);
+    u32 res = this->cyl.ChkTgHit();
+    if (res != 0) {
+        setSpeed(*this->cyl.GetTgRVecP());
+    }
+
+    setMtx();
+    return TRUE;
 }
 
 /* 00000884-00000888       .text executeNormal__10daObjHat_cFv */
-void daObjHat_c::executeNormal() {
-    /* Nonmatching */
-}
+void daObjHat_c::executeNormal() { /* empty? */ }
 
 /* 00000888-000008B4       .text getPrmHatNo__10daObjHat_cFv */
-void daObjHat_c::getPrmHatNo() {
-    /* Nonmatching */
+u32 daObjHat_c::getPrmHatNo() {
+    return daObj::PrmAbstract(this, PRM_SWSAVE_W, PRM_SWSAVE_S) & 3;
 }
 
 /* 000008B4-00000964       .text setMtx__10daObjHat_cFv */
 void daObjHat_c::setMtx() {
-    /* Nonmatching */
+    model->mBaseScale = this->scale;
+    MTXTrans(mDoMtx_stack_c::now, this->current.pos.x, this->current.pos.y,
+             this->current.pos.z);
+    mDoMtx_YrotM(mDoMtx_stack_c::now, this->current.angle.y + 0x4000);
+    mDoMtx_XrotM(mDoMtx_stack_c::now, this->current.angle.x);
+    mDoMtx_ZrotM(mDoMtx_stack_c::now, this->current.angle.z + 0x4000);
+    MTXCopy(mDoMtx_stack_c::now, this->model->mBaseTransformMtx);
 }
 
 /* 00000964-00000A20       .text setSpeed__10daObjHat_cF4cXyz */
-void daObjHat_c::setSpeed(cXyz) {
-    /* Nonmatching */
+void daObjHat_c::setSpeed(cXyz speed) {
+    speed.y = 0;
+    speed = speed.normZP();
+
+    this->angle = speed;
+
+    fopAcM_SetSpeedF(this, 20.0f);
+
+    speed.x *= 20.0f;
+    speed.y = 50.0f;
+    speed.z *= 20.0f;
+
+    fopAcM_SetSpeed(this, speed.x, speed.y, speed.z);
 }
 
 /* 00000A20-00000A40       .text daSampleCreate__FPv */
-static s32 daSampleCreate(void*) {
-    /* Nonmatching */
+static s32 daSampleCreate(void* arg) {
```

Comment:
```suggestion
static s32 daSampleCreate(void* i_this) {
```

The parameter holding the current actor is pretty much always named `i_this`. Same suggestion applies to other funcs